pkg/cli/completion.go

```diff
@@ -0,0 +1,257 @@
+package cli
+
+import (
+	"context"
+	"strings"
+
+	"github.com/acorn-io/acorn/pkg/client"
+	"github.com/acorn-io/baaah/pkg/typed"
+	"github.com/spf13/cobra"
+	"k8s.io/utils/strings/slices"
+)
+
+type completionFunc func(context.Context, client.Client, string) ([]string, error)
+
+// These define instances when the completion should not happen and the default completion (usually file/directory
+// completion should occur). For example, exec should only have one argument, everything after that is completed as
+// default by the user's terminal.
+type noCompletionOption func([]string) bool
+
+type completion struct {
+	client           client.ClientFactory
+	completionFunc   completionFunc
+	successDirective cobra.ShellCompDirective
+
+	noCompletionOptions []noCompletionOption
+}
+
+func removeExistingArgs(result, args []string) []string {
+	for i := 0; i < len(result); {
+		if slices.Contains(args, result[i]) {
+			result = append(result[:i], result[i+1:]...)
+		} else {
+			i++
+		}
+	}
+
+	return result
+}
+
+func onlyNumArgs(n int) noCompletionOption {
+	return func(args []string) bool {
+		return len(args) >= n
+	}
+}
+
+func newCompletion(c client.ClientFactory, cf completionFunc) *completion {
+	return &completion{
+		client:           c,
+		completionFunc:   cf,
+		successDirective: cobra.ShellCompDirectiveNoFileComp,
+	}
+}
+
+func (a *completion) withShouldCompleteOptions(opts ...noCompletionOption) *completion {
+	a.noCompletionOptions = append(a.noCompletionOptions, opts...)
+	return a
+}
+
+func (a *completion) withSuccessDirective(d cobra.ShellCompDirective) *completion {
+	a.successDirective = d
+	return a
+}
+
+func (a *completion) complete(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	for _, o := range a.noCompletionOptions {
+		if o(args) {
+			return nil, cobra.ShellCompDirectiveDefault
+		}
+	}
+	c, err := a.client.CreateDefault()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	result, err := a.completionFunc(cmd.Context(), c, toComplete)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	return removeExistingArgs(result, args), a.successDirective
+}
+
+func appsThenContainersCompletion(ctx context.Context, c client.Client, toComplete string) ([]string, error) {
```

Comment:
Thoughts on moving these completion functions out of here and into their relevant `cli/<command>.go files`? This would make it so that this file doesn't expand too much overtime and keep more relevant code together. Then the other boilerplate settings (like `withSuccessDirective` and `completion`) would stay here.

Reply:
The main reason I didn't do that is that some of these are used in multiple places. For example, the apps completion is used in a number of commands (apps, start, stop, rm, etc).

Secondly, these functions can be unit tested independent of the commands themselves, which is how I implemented the tests.

I'm open to moving them though. After reading my explanation, do you still feel like moving them would be better?

Reply:
Nope! I'm with you on the benefits of it being like it is now.

Reply:
Awesome! Thanks for the review, @tylerslaton!